packages/backend/src/payment-method/ilp/connector/core/middleware/balance.ts

```diff
@@ -21,96 +21,103 @@ export function createBalanceMiddleware(): ILPMiddleware {
+            amount,
+            destinationAmountOrError,
+            sourceAsset: accounts.incoming.asset,
+            destinationAsset: accounts.outgoing.asset
+          },
+          'Could not get rates'
         )
-        switch (trxOrError) {
-          case TransferError.InsufficientBalance:
-          case TransferError.InsufficientLiquidity:
-            throw new InsufficientLiquidityError(trxOrError)
-          default:
-            // TODO: map transfer errors to ILP errors
-            ctxThrow(500, destinationAmountOrError.toString())
+        throw new CannotReceiveError(
+          `Exchange rate error: ${destinationAmountOrError}`
+        )
+      }
+      const { amount: destinationAmount } = destinationAmountOrError
+
+      request.prepare.amount = destinationAmount.toString()
+
+      if (state.unfulfillable) {
+        await next()
+        return
+      }
+
+      // Update balances on prepare:
+      const createPendingTransfer = async (): Promise<
+        Transaction | undefined
+      > => {
+        const transferOptions = {
+          sourceAccount: accounts.incoming,
+          destinationAccount: accounts.outgoing,
+          sourceAmount,
+          destinationAmount,
+          transferType: TransferType.TRANSFER,
+          timeout: AppConfig.tigerBeetleTwoPhaseTimeout
+        }
+        const trxOrError =
+          await services.accounting.createTransfer(transferOptions)
+        if (isTransferError(trxOrError)) {
+          logger.error(
+            { transferOptions, transferError: trxOrError },
+            'Could not create transfer'
+          )
+          switch (trxOrError) {
+            case TransferError.InsufficientBalance:
+            case TransferError.InsufficientLiquidity:
+              throw new InsufficientLiquidityError(trxOrError)
+            default:
+              // TODO: map transfer errors to ILP errors
+              ctxThrow(500, destinationAmountOrError.toString())
+          }
+        } else {
+          return trxOrError
         }
-      } else {
-        return trxOrError
       }
-    }
 
-    if (state.streamDestination) await next()
+      if (state.streamDestination) await next()
 
-    if (!state.streamDestination || response.fulfill) {
-      // TODO: make this single-phase if streamDestination === true
-      const trx = await createPendingTransfer()
+      if (!state.streamDestination || response.fulfill) {
+        // TODO: make this single-phase if streamDestination === true
+        const trx = await createPendingTransfer()
 
-      if (!state.streamDestination) await next()
+        if (!state.streamDestination) await next()
 
-      if (trx) {
-        if (response.fulfill) {
-          await trx.post()
-        } else {
-          await trx.void()
+        if (trx) {
+          if (response.fulfill) {
+            await trx.post()
+          } else {
+            await trx.void()
+          }
         }
       }
+    } finally {
+      stopTimer()
```

Comment:
unfortunately, I think this timer will encapsulate all of the following middlewares in the timing (because we will be waiting for all the `await next()`s.). We can tackle this in a separate PR, we could probably do something nice in the `compose` function in `src/payment-method/ilp/connector/index.ts` to automatically add telemetry for all middlewares

Reply:
(same for the error handler middleware)